### ISSUE TITLE
[5] Test-time Robust Personalization for Federated Learning

### ISSUE THREAD

**Issue by ejjun92** [ejjun92/Autonomous-Driving-Papers#5]
<img width="1405" alt="스크린샷 2023-06-27 오후 9 28 49" src="https://github.com/ejjun92/Autonomous-Driving-Papers/assets/26243545/28fad177-2cca-4f00-9462-2c0e58dd6a8c">

* Problem: real-world에서 deployment 동안 distribution shift 발생하는 문제에 대해
여전히 FL 모델이 robust 할 수 있도록
* Proposed: 다양한 test-time distribution shift에 대한 robust한 personalized FL 모델 제공
Federated Test-time Head Ensemble plus tuning (FedTHE+)
* Evaluation: CIFAR10 & ImageNet에서
다양한 NN 아키텍쳐 (CNN, ResNet, Transformer) 학습하여 
FedTHE+ / FED 평가

## Background
* Federated learning (FL): 여러 client들이 분산 data를 갖고 공동으로 shared global model을 학습하는 ML 알고리즘
* Personalized FL (PFL): global model을 각 client에 맞게 adapt시킴.
In-Distribution (ID)에서 좋은 결과 보여주었으나
  1) 입력되는 local test sample이 계속해서 바뀌고,
  2) 다양한 type의 OOD shift가 발생하기 때문에 
실제 FL deployment가 어려움

<img width="1405" alt="스크린샷 2023-06-27 오후 9 29 00" src="https://github.com/ejjun92/Autonomous-Driving-Papers/assets/26243545/33c6422d-5050-4a3b-b3da-1018989c0afc">

* 가능한 test-time distribution shift 시나리오
  * label distribution shift
    * client들이 보지 못한 class의 새로운 local test 샘플 
    * 새로운 class가 나타나지 않더라도, local test 분포가 다른 경우
  * co-variate shift
    * 또다른 real-world distribution shift (synthetic distribution shift or natural distribution shift)
 
* FL을 practical하게 하기 위해
  1) FL 모델을 ID & OOD 모두에서 일반화하고
  2) deployment robustness를 적절히 평가하는게 중요함
* 이를 위해, client들이 배포 중에 직면하는 ID & OOD 상황을 모방하는 새로운 PFL 벤치마크 구축 

* Test Time Adaptation (TTA)
  * OOD generalization을 위한 training 전략과는 별개로
test-time distribution shift를 완화하는데 굿.
* 근데 현재의 TTA 방식은 homogeneous & non-distributed 환경에서 설계되어
FL distribution shift 시나리오에서는 제한적인 성능을 보임.

### PFL
* 가장 직관적인 방법은 global 모델을 locally fine-tuning.
* locally 학습한 모델과 global 모델을 linearly interpolate하여 평가.
* 여러 local model들에 대해 weight combination.
* NN을 shared feature extractor와 personalized head로 decompose.
* FedRoD: 2 head network / local & global optimization objective 분리
* pFedHN & ITU-PFL: hypernetwork 사용
* ITU-PFL은 레이블이 없는 새로운 client 이슈에 초점을 맞춤.

-> limitation: 기존 PFL 방법들은 local ID 상황의 일반화에 초점을 맞추어, 
test-time distribution shift의 OOD robustness가 부족함.

### FL에서 OOD generalization
* 기존 방법들은 단지 분포적으로 robust한 global model을 optimize하고, 높은 test-time accuracy 달성 X
* GMA: client들에 대한 invariant 메커니즘 학습을 위해 client update에 masking 전략 사용.
* Fairness-aware FL도 유사한 컨셉 사용.
* global model의 정확도 안정성은 local training distirubution 전반에 걸쳐 강화됨.

-> 그러나, real-world common corruptions, natural distribution shifts, label distribution shifts에 대해서 연구되지 X.

### FL 벤치마킹
* 기존의 FL 벤치마크들은 PFL의 다양한 ID & OOD 케이스들에 대한 test accuracy를 적절하게 측정 X.
* 제안하는 Benchmark for Robust FL (BRFL)은 
covariate & label distribution shift로 구성된 시나리오를 다루어 이러한 간극 메움.

### Test-time Adaptation (TTA)
* 초기에 OOD 일반화를 위해 Test-Time Training (TTT) 에서 제안됨.

* TTA는 2-head NN 구조를 사용했는데,
여기서 self-supervised auxiliary head/task (rotation prediction)가 feature extractor를 fine-tuning하기 위해 사용됨.

* TTT++는 여기에 feature alignment block을 추가함.
그러나, 테스트 하기 전에 전체 테스트 데이터셋에 access 해야하는 점 때문에 online deployment 동안 실행하기에 어려움.

* Tent는 prediction entropy를 최소화하고, 오직 Batch normalization 파라미터들만 update.
* MEMO는 augmented views에 대한 marginal entropy를 최소화함으로써 모든 모델 파라미터들을 adapt.
* TSA는 augmented views 간의 prediction similarity를 최대화하여 long-tailed recognition 개선.

* T3A는 학습된 linear classifier를 class prototype으로 대체하고,
각 sample을 prototype에 대한 거리를 기반으로 분류함.

* CoTTA는 class-balanced test sample이 순차적으로 다양한 corruption type을 직면하는 continual 시나리오를 고려.

-> 제안 방법은 global & local objective로 학습한 2 FL-specific head 설계.
optimal head interpolation weight 학습.

## ID & OOD에 대한 adaptation 전략
* 분포에 대한 pre-trained model을 optimize하고 
ID or OOD 데이터에서 샘플링된 unknown test sample에 대해 평가할 때,
adaptation 전략의 선택에서 성능 상 trade-off 존재.
* 제안 방법은 ID 정확도를 희생하지 않고 OOD 정확도 개선을 목표로 함.

* 2-stage adaptation 전략
1) Fine-Tuning (FT): 모든 모델 파라미터들을 업데이트 (ID에 효과적)
2) Linear Probing (LP): 마지막 linear layer (head)만 업데이트

* FT / LP: ID / OOD 데이터 간 test performance에 있어 trade-off
* B: feature extractor, v: head 일 때,
* B_0 -> B_*: LP는 near-optimal linear head 학습 -> small OOD error
* FT는 v_FT와 B_FT 간 coupled gradient update로 인해 high OOD error 
* v_0와 v_* 간 distance는 B_FT를 shift -> feature를 distort -> higher OOD error 
* 만약 v가 v_*라면, FT update가 OOD error를 높히지 않을 것

-> 먼저 LP하고, 그 다음에 FT하자! (LP-FT)

## FedTHE+
* 기능: FL 모델을 personalize & robustify
* 방법:
    * single test-sample에 대해
global generic & local personalized classifier를
unsupervised & adaptively ensemble
    * unsupervised test-time fine-tuning 
* contribution
    * acamia의 현재 PFL 방법들과 real-world deployment 니즈 간 gap: test-time distribution shift 이슈 확인 및 PFL 재평가
    * baselines, NNs, datasets, test distribution shifts에 대해 robust한 ID &OOD 성능을 보이는 PFL 프레임워크 제안

### Test-time Robust Personalization for Federated Learning

supervised 2-stage LP-FT -> unsupervised test-time PFL로 확장

1) Unsupervised Linear Probing (FedTHE): single test sample에 대해 scalar head ensemble weight를 효율적으로 최적화하여 훨씬 더 robust한 예측 생성
2) Unsupervised Fine-tuning: 1)에 unsupervised FT (MEMO) 수행하여 더 높은 정확도와 robustness 얻음.

<img width="1414" alt="스크린샷 2023-06-27 오후 9 29 19" src="https://github.com/ejjun92/Autonomous-Driving-Papers/assets/26243545/36fda152-8aee-4314-ad7d-6014653b3749">
<img width="1362" alt="스크린샷 2023-06-27 오후 9 29 36" src="https://github.com/ejjun92/Autonomous-Driving-Papers/assets/26243545/2b229b5f-00a4-48c6-9630-aca8f557e9bc">

* standard FLL global objective function은 K개의 client들의 local objective function들의 weighted sum

* traditional FL: heterogeneous local data 분포들에 대해 optimal global model을 optimize하는 것이 목적
* personalized FL: 각 local data 분포에 collaboratively & individually adaptation -> relaxed optimization objective.
즉, client-wise personalized model이 locally 계산됨.

* linear probing으로써 FedTHE
  * 기존의 PFL은 per-client personalized model을 각 local training 분포에 adapt 시킴.
-> 결과, PFL 모델은 레이블 분포 변화에 robustness을 잃게됨.

* global 모델은 모든 클래스 레이블에 대해 일반화 되지만,
* personalized 모델은 biased class 분포를 학습하게 될 수 있고,
FL동안 학습된 global 모델의 invariance가 naive adaptation에 의해 방해받음.
* global 모델의 robustness를 보존 + local distribution에서 좋은 성능 달성 ?

* 2-head FL 네트워크: feature extractor + global head + personalized head

* 알고리즘
1) Federated Training
기존 FedAvg와 유사하게,
각 client는 local training 데이터로 받은 global model을 학습시킴.
-> global model을 서버로 다시 보냄.
-> personalized head는 locally 학습되고 유지됨. (frozen & unadapted feature extractor)

2) Test-time Robust Personalization
deployment동안 ID & OOD 다루기 위해,
single test sample에 대해 
e를 optimize하면서 personalized head & global head를 interpolate (unsupervised linear probing)

* prediction uncertainty 줄이기
  * 2-head weighted logits prediction의 Shannon Entropy 줄임.
  *e를 optimize 하는건 prediction confidence를 향상시킴.
  * 그러나 1) EM을 통한 높은 prediction confidence는 항상 accuracy의 향상은 아님,
2) local training 데이터로 학습된 personalized head는
biased class 분포에 쉽게 overfit되며,
이러한 overconfidence는 arbitrary distribution shift에 adaptation quality와 reliability를 해침.

* feature space alignment
    * EM의 limitation에 대응하기 위해, 
feature space constraint 적용하여 head interpolation weight를 optimize.
    * Feature Alignment (FA) loss를 줄이는 것은
test feature에 더 가까운 head (local personalized or global generic)에 더 높은 weight 할당

Feature descriptor (per-client local descriptor / global descriptor / test history descriptor)
모두 server의 global feature extractor로부터 계산됨.

  (1) local descriptor는 local training sample들의 feature representation의 평균. 

  (2) federated learning동안 각 client는 학습된 global model parameter와 함께 자신의 local descriptor를 서버로 전달하고, 서버는 모든 local descriptor를 평균하여 global descriptor 생성.
sampled clients에 global descriptor, global model 동기화됨.

  (3) single test feature 내 high variance의 instability를 줄이기 위해, 
Exponential Moving Average (EMA)를 통한 test history decriptor 사용.

* Similarity-guided Loss Weighting (SLW)
    * global & personalized head로부터 probability output의 cosine similarity를 갖는
adaptive unsupervised loss 사용.
    * 즉, 2 head로부터 유사한 logit prediction 일 때 (높은 cosine similarity)
more confident head 선호되고,
유사하지 않은 logit prediction 일 때 (낮은 cosine similarity)
feature alignment loss가 overfitting issue를 regularize
& prediction confidence와 head expertise 간에 equilibrium 달성.

* test-time FT 단계에 MEMO (Zhang et al., 2021a) 사용. -> FedTHE+
* test-time FT는 optimized head ensemble weight를 고정한 상태에서 
네트워크 매개변수 (w_e, w_g, w_l)를 업데이트. 
* FedTHE를 적용하는 것만으로도 다양한 baselines과 비교하여 향상된 ID 및 OOD 정확도를 얻는 데 충분하며, FedTHE+는 더 높은 정확도와 robustness 향상을 달성하며 LP-FT의 이점을 얻음.

### BRFL: Robust FL을 위한 벤치마크
1) disjoint non-iid 데이터를 가진 heterogeneus clients 생성하기 위해
clents끼리 shuffle되지 않도록 Dirichlet 분포 사용.
heterogeneity 정도가 작을수록 client가 적은 클래스 예제를 보유할 확률 높음.

2) 각 client의 데이터를 local training/validation/test 세트로 균일하게 분할.

* 다양한 테스트 분포 -> 5개의 distinct 테스트 분포: 1 ID & 4 OODs
  * ID: 원래 local test
  * OOD
    * corrupted local test, naturally shifted test (covariate shift)
    *  Out-of-Client (OoC) local test (label distribution shift)
    * Mixture of test
